safemysql.class.php

```diff
@@ -517,35 +554,66 @@ public function getStats()
 	/**
 	 * private function which actually executes $query.
 	 * also logs some stats like profiling info and error message
-	 * 
+	 *
 	 * @param string $query - a regular SQL query
 	 * @return mysqli result resource or FALSE on error
 	 */
 	private function rawQuery($query)
 	{
-		$start = microtime(TRUE);
-		$res   = mysqli_query($this->conn, $query);
-		$timer = microtime(TRUE) - $start;
+		return $this->retryIfDeadlocked(
+			function($db) use($query)
+			{
+				$start = microtime(TRUE);
+				$res   = mysqli_query($this->conn, $query);
+				$timer = microtime(TRUE) - $start;
+
+				$this->stats[] = array(
+					'query' => $query,
+					'start' => $start,
+					'timer' => $timer,
+				);
+				if (!$res)
+				{
+					$error = mysqli_error($this->conn);
+					$errno = mysqli_errno($this->conn);
 
-		$this->stats[] = array(
-			'query' => $query,
-			'start' => $start,
-			'timer' => $timer,
+					end($this->stats);
+					$key = key($this->stats);
+					$this->stats[$key]['error'] = $error;
+					$this->cutStats();
+
+					$this->error("$error. Full query: [$query]", $errno);
+				}
+				$this->cutStats();
+				return $res;
+			}
 		);
-		if (!$res)
+	}
+
+	private function retryIfDeadlocked($callback, $can_retry = null)
+	{
+		if ($can_retry === null)
 		{
-			$error = mysqli_error($this->conn);
-			$errno = mysqli_errno($this->conn);
-			
-			end($this->stats);
-			$key = key($this->stats);
-			$this->stats[$key]['error'] = $error;
-			$this->cutStats();
-			
-			$this->error("$error. Full query: [$query]", $errno);
+			$can_retry = $this->retryOnDeadlock;
+		}
+		for ($i = 0; $i <= $this->maximumRetriesOnDeadlock; $i++)
+		{
+			try {
+				return $callback($this);
+			} catch (Exception $e)
+			{
+				// Only retry if this is a deadlock, this callable is allowed to be retried, and we have not yet reached
+				// the maximum number of retries
+				if ($can_retry && $e->getCode() === 1213 && $i < $this->maximumRetriesOnDeadlock)
+				{
+					sleep(2 ** $i);
```

Comment:
- [ ] Again, this isn't compatible with the version declared in composer.json, at least according to http://stackoverflow.com/a/22038259/1709587 which says that `**` was only introduced in PHP 5.6 (haven't checked docs to confirm).